Mortality

mcwimm opened the pull request:
This is a selection of Mortality concepts applicable to SimpleBettina and NetworkBettina (see docs).

This should not affect the functionality of previous scripts. If no mortality-concept is chosen, trees die immediatly if grow < 0 (as currently in the master).

jbathmann commented on TreeOutputLib/OneFile/OneFile.py:
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
```
unrelated. please remove from this pr

mcwimm replied:
solved

jbathmann commented on TreeOutputLib/OneTimestepOneFile/OneTimestepOneFile.py:
```diff
@@ -63,6 +63,8 @@ def __init__(self, args):
             " and geometric" + " measures ", self.geometry_outputs,
```
unrelated. please remove from this pr

mcwimm replied:
solved

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:
```diff
@@ -154,7 +162,7 @@ def getInputParameters(self, args):
             for tag in missing_tags:
                 string += tag + " "
-            raise KeyError(
+            print(
```
see previous comment

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/SimpleBettina/SimpleBettina.py:
```diff
@@ -178,6 +194,41 @@ def growthResources(self):
-        if (self.grow < 0):
-            self.grow = 0
+        # Iterate through mortality concepts
+        # if tree dies in one of them he's dead
+        survive = []
+        for mortality_concept in self.mortality_concept:
+            survive.append(mortality_concept.getSurvival(self))
+
+        if 0 in survive:
             self.survive = 0
+
+    def getMortalityVariables(self, growth_concept_information):
+        if "Memory" in self.mortality_concept_names:
+            try:
+                self.grow = growth_concept_information["growth"]
+                self.grow_memory = growth_concept_information["grow_memory"]
+            except KeyError:
+                self.grow = 0
+                self.grow_memory = []
+            self.grow_before = self.grow
+
+        if "MinGirthGrowth" in self.mortality_concept_names:
+            try:
+                self.r_stem_memory = growth_concept_information["r_stem_memory"]
+            except:
+                self.r_stem_memory = []
+            self.r_stem_before = self.r_stem
+
+    def setMortalityVariables(self, growth_concept_information):
+        if "Memory" in self.mortality_concept_names:
+            growth_inc = self.grow - self.grow_before
+            self.grow_memory.append(growth_inc)
+            growth_concept_information["grow_memory"] = \
+                self.grow_memory
+        if "MinGirthGrowth" in self.mortality_concept_names:
+            r_stem_inc = self.r_stem - self.r_stem_before
+            self.r_stem_memory.append(r_stem_inc)
+            growth_concept_information["r_stem_memory"] = \
+                self.r_stem_memory
+        return growth_concept_information
```
can you do this in a "cleaner" way? Maybe by adding it to the normal GrowthConcept-class?

mcwimm replied:
Solved

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/Mortality/modules/MinGirthGrowth.py:
```diff
@@ -0,0 +1,69 @@
+              str(self.min_girth_growth) + " cm and memory of " +
+              str(self.period) + " years.")
+
+    def getSurvival(self, args):
+        self.survive = 1
+
+        # get the number of values representing the memory period
+        period_in_sec = self.period * 3600 * 24 * 365
+        steps = int(period_in_sec / args.time)
+
+        # slice grow_memory array to get only relevant data
+        r_stem_inc = args.r_stem_memory[-int(steps):]
+
+        # exclude the first time steps as the stem does not grow in the
+        # beginning in BETTINA
+        if len(r_stem_inc) >= int(steps):
+            # calculate current growth increment
+            avg_girth_growth = np.mean(r_stem_inc)
+            steps_per_year = (3600 * 24 * 365) / args.time
+            avg_girth_growth_y = avg_girth_growth * steps_per_year
+
+            # transform yearly DBH increment rate in r_stem in m
+            min_girth_growth_y_m = self.min_girth_growth / 2 / 100   # in m
+            if avg_girth_growth_y < min_girth_growth_y_m:
+                self.survive = 0
+                print("\t Tree died because average DBH increment fell below "
+                      "threshold.")
```
This is repeated from other classes... inheritance?

mcwimm replied:
I think it's not repeated as each mortality class contains another approach to calculate the survival factor in `getSurvival`.

jbathmann replied:
and there is also no partwise repeating?

mcwimm replied:
Not any more because there is a new concept (after discussion with @zaphod-bx ) and others (Memory, MinGirthGrowth) were removed.

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/Mortality/modules/Memory.py:
```diff
@@ -0,0 +1,67 @@
+    def getSurvival(self, args):
+        self.survive = 1
+        # get the number of values representing the memory period
+        period_in_sec = self.period * 3600 * 24 * 365
```
what are you ding here? the input should be in SI units, as we are using SI throughout the model

mcwimm replied:
changed :)

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/Mortality/modules/MinGirthGrowth.py:
```diff
@@ -0,0 +1,69 @@
+
+class MinGirthGrowth:
+    def __init__(self, args, case):
+        self.survive = 1
+        self.min_girth_growth = 0.05   # cm per year (Grueters et al. 2021)
+        self.period = 5
+        self.getMinGrowthValue(args)
+
+        print("Initiate mortality of type " + case +
+              " with min. yearly DBH growth of " +
+              str(self.min_girth_growth) + " cm and memory of " +
+              str(self.period) + " years.")
```
please inherit repeated content

mcwimm replied:
parameters differ between concepts

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/Mortality/modules/Random.py:
```diff
@@ -0,0 +1,50 @@
+        self.survive = 1
+        r = np.random.random(1)
+        # Number of time steps per year
+        steps_per_year = (3600 * 24 * 365) / args.time
```
??
A year has 365.25 days. ;-)

mcwimm replied:
right :) changed

jbathmann reviewed (commented):
I like, where this is going!

mcwimm commented on TreeOutputLib/OneTimestepOneFilePerGroup/OneTimestepOneFilePerGroup.py:
```diff
@@ -28,14 +28,8 @@ def writeOutput(self, tree_groups, time):
                 string = ""
```
solved

mcwimm commented on TreeOutputLib/__init__.py:
```diff
@@ -10,4 +10,5 @@
 from .TreeOutput import TreeOutput
```
solved

mcwimm commented on Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/Mortality/c_Mortality.md:
```diff
@@ -0,0 +1,56 @@
+### Concept `NoGrowth`
+
+In `NoGrowth`, a tree dies if the growth variable is equal or below 0, i.e. no growth happens. 
+
+This concept is based on mechanistic causes, e.g. resource deficit, and no tree dies of random reasons.
+
+
+### Concept `Random`
+
+In `Random`, a tree dies with a defined yearly mortality probability.
+If a random number times the number of time steps per year is below this defined probability the tree dies.
+The default probability is 0.2 %. Berger & Hildenbrandt (2000) suggested 0.2 and 0.16 % for Rhizophora mangle and Avicennia germinans, respectively.  
+Use `<probability> 0.002 </probability>` to modify concept parameters.
+
+In this concept, a tree does not die of mechanistic reasons. It is useful to combine `Random` with `NoGrowth` or `Memory`.  
+
+
+### Concept `Memory`
+
+In `Memory`, a tree remembers its biomass increment of the preceding N years.
+If the current biomass increment is below a certain percentage of average growth increment of the last N years the tree dies (similar to Berger & Hildenbrandt (2000)).
+The period the memory lasts (N) is defined with `<period> </period>` in the xml file (default: 5 years).
+The threshold (percentage) is defined with `<threshold> </threshold>` in the xml file (default: 0.5). The value must be a number between 0 and 1.  
+Use `<period> 2 </period>` and `<threshold> 0.5 </threshold>` to modify concept parameters.
+
+This concept is based on mechanistic causes, e.g. slowing down of growth due to tree age, and no tree dies of random reasons.
+
+
+### Concept `MinGirthGrowth`
+
+The concept is similar to `Memory` but uses girth growth as the relevant parameter to assess survival.  
+
+`MinGirthGrowth` requires a minimum yearly girth growth rate in [cm per year]. 
+If a trees moving average girth growth over the previous N years fell below the minimum rate, the tree dies.
+The default growth rate is 0.05 cm per year (Grueters et al. 2021 for Rhizophora).  
+Use `<period> 2 </period>` and `<min_girth_growth> 0.5 </min_girth_growth>` to modify concept parameters.
+
+NOTE: The concept is only applied to trees with age > N as otherwise trees will die immediately.  
+That is, the tree dies not immediately after recruitment.
+
```
This fits to the current implementation as the description of concepts here is just a explanation for possible inputs for the tag `mortality`. Nonetheless, this might change if we restructure the mortality stuff...

jbathmann replied:
ok. how do you proceed wth this?

mcwimm replied:
Solved

mcwimm commented on TreeOutputLib/OneTreeOneFile/OneTreeOneFile.py:
```diff
@@ -13,16 +13,16 @@
 #  A line containing time, position, desired geometric measures and desired
```
solved

mcwimm commented on TreeOutputLib/TreeOutput.py:
```diff
@@ -21,8 +21,11 @@ def __init__(self, args):
             self.iniOneTimestepOneFile(args)
```
solved

mcwimm commented on Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/Mortality/c_Mortality.md:
```diff
@@ -0,0 +1,56 @@
+
+In `Memory`, a tree remembers its biomass increment of the preceding N years.
+If the current biomass increment is below a certain percentage of average growth increment of the last N years the tree dies (similar to Berger & Hildenbrandt (2000)).
+The period the memory lasts (N) is defined with `<period> </period>` in the xml file (default: 5 years).
+The threshold (percentage) is defined with `<threshold> </threshold>` in the xml file (default: 0.5). The value must be a number between 0 and 1.  
+Use `<period> 2 </period>` and `<threshold> 0.5 </threshold>` to modify concept parameters.
```
I would prefer default inputs as there's not much information available on these parameters. I presume that we'll usually use the default values.

jbathmann replied:
ok...

mcwimm commented on Documentation/project_dox/MangaProject/tree_dynamics/tree_growth_and_death/t_type.dox:
```diff
@@ -7,6 +7,7 @@ Case selector for different types of aboveground competition.
 
 # Available cases 
  
+- Mortality
```
At the moment, mortality parameters are additional/ optional parameters for growth-and-death-concepts and they are defined as follows:

```
<tree_growth_and_death>  
     <type> SimpleBettina</type>  
     <mortality> Memory </mortality>  
     <period> 1e9  </period>  
     <threshold> 0.5 </threshold>
</tree_growth_and_death>
```

jbathmann replied:
<tree_growth_and_death>  
     <type> SimpleBettina</type>  
     <mortality>
        <type> Memory </type> 
        <period> 1e9  </period>  
         <threshold> 0.5 </threshold>
     </mortality>  
</tree_growth_and_death>

jbathmann replied:
is it possible to follow this structure? (no idea how to format correctly here... -.-)

mcwimm commented on TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:
```diff
@@ -137,7 +144,8 @@ def rootGraftFormationV2(self):
 
     def getInputParameters(self, args):
-        missing_tags = ["type", "variant", "f_growth"]
+        missing_tags = ["type", "variant", "f_growth",
+                        "mortality", "threshold", "period"]
```
No, it's an optional input. If it is not defined the mortality concept 'NoGrowth' is used (this is the current mortality concept in the master).

jbathmann replied:
but missing_tags is listing the required inputs (see line 165)

mcwimm commented:
@jbathmann I changed three major thinks with the last commits:

- I deleted the former concepts ('Memory' and 'MinGirthGrowth') after a discussion with @zaphod-bx about their usefullness.
- The mortality concept 'NoGrowth' is now the super class for all other mortality concepts, containing all required methods.
- I restructured the Mortality directory in order to align the structure to other sections such as belowground competition. However, that is not the structure you suggested... The mortality concept is still initialized by the growth concept.

jbathmann commented:
For me it is ready to merge...

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:
```diff
@@ -19,6 +20,12 @@ def __init__(self, args):
         print("Growth and death dynamics of type " + case + ".")
         self.getInputParameters(args)
 
+        M = Mortality.Mortality(args)
+        self.mortality_concept = M.getMortConcept()
+        self.mortality_concept_names = []
+        for concept in self.mortality_concept:
+            self.mortality_concept_names.append(concept.getConceptName())
+
```
please move this to super()

mcwimm replied:
solved.

jbathmann reviewed (changes requested):
Please add the mortality stuff to the superclass (treemodel) and than just call it in the constructor of your individual concepts, if need be.

jbathmann commented on TreeModelLib/GrowthAndDeathDynamics/SimpleBettina/SimpleBettina.py:
```diff
@@ -178,6 +199,11 @@ def growthResources(self):
         self.grow = (self.parameter["growth_factor"] *
                      (self.available_resources - self.maint))
-        if (self.grow < 0):
-            self.grow = 0
+        # Iterate through mortality concepts
+        # if tree dies in one of them he's dead
+        survive = []
+        for mortality_concept in self.mortality_concept:
+            survive.append(mortality_concept.getSurvival(self))
+
+        if 0 in survive:
             self.survive = 0
```
why here?!

jbathmann replied:
nevermind

mcwimm commented:
> Please add the mortality stuff to the superclass (treemodel) and than just call it in the constructor of your individual concepts, if need be.

Done.

jbathmann commented:
looks ready to me.
